src/main/java/de/tum/in/www1/artemis/service/AchievementService.java

```diff
@@ -0,0 +1,78 @@
+package de.tum.in.www1.artemis.service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.Achievement;
+import de.tum.in.www1.artemis.domain.Course;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.enumeration.AchievementRank;
+import de.tum.in.www1.artemis.repository.AchievementRepository;
+import de.tum.in.www1.artemis.repository.UserRepository;
+
+@Service
+public class AchievementService {
+
+    private final AchievementRepository achievementRepository;
+
+    private final UserRepository userRepository;
+
+    public AchievementService(AchievementRepository achievementRepository, UserRepository userRepository) {
+        this.achievementRepository = achievementRepository;
+        this.userRepository = userRepository;
+    }
+
+    public Optional<Achievement> findById(Long achievementId) {
+        return achievementRepository.findById(achievementId);
+    }
+
+    public List<Achievement> findAll() {
+        return achievementRepository.findAll();
+    }
+
+    public Set<Achievement> findAllByCourseId(Long courseId) {
+        return achievementRepository.findAllByCourseId(courseId);
+    }
+
+    public Set<Achievement> findAllByExerciseId(Long exerciseId) {
+        return achievementRepository.findAllByExerciseId(exerciseId);
+    }
+
+    public Set<Achievement> findAllByUserId(Long userId) {
+        return achievementRepository.findAllByUserId(userId);
+    }
+
+    public Achievement save(Achievement achievement) {
+        return achievementRepository.save(achievement);
+    }
+
+    /**
+     * Creates an achievement and persist it
+     */
+    public Achievement create(String title, String description, String icon, AchievementRank rank, Course course) {
```

Comment:
Don't you need to add a user relationship somewhere, when you give out achievements?

Reply:
no, not on creation itself, only on assignment of the achievement later

Reply:
Should also have a connection to an exercise

Reply:
I think this is ok. You can create the achievement first and then add the relationship.

Reply:
I guess you can add the connection to an exercise later, but it could also be possible right here.